plugins/haconfig/keepalived/generator.go

```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		glog.Fatalf("Credentials %q error: %v", path, err)
+	}
+
+	if err := kclient.LoadTLSFiles(config); err != nil {
+		glog.Fatalf("Unable to load certificate info using credentials from %q: %v", path, err)
+	}
+
+	return config
+}
+
+//  Generate the HA failover monitor (keepalived) container environment entries.
+func generateEnvEntries(name string, options *haconfig.HAConfigCmdOptions, kconfig *kclient.Config) app.Environment {
+	watchPort := strconv.Itoa(options.WatchPort)
+	replicas := strconv.Itoa(options.Replicas)
+	insecureStr := strconv.FormatBool(kconfig.Insecure)
+	unicastStr := strconv.FormatBool(options.UseUnicast)
+
+	return app.Environment{
+		"OPENSHIFT_MASTER":    kconfig.Host,
+		"OPENSHIFT_CA_DATA":   string(kconfig.CAData),
+		"OPENSHIFT_KEY_DATA":  string(kconfig.KeyData),
+		"OPENSHIFT_CERT_DATA": string(kconfig.CertData),
+		"OPENSHIFT_INSECURE":  insecureStr,
+
+		"OPENSHIFT_HA_CONFIG_NAME":       name,
+		"OPENSHIFT_HA_VIRTUAL_IPS":       options.VirtualIPs,
+		"OPENSHIFT_HA_NETWORK_INTERFACE": options.NetworkInterface,
+		"OPENSHIFT_HA_MONITOR_PORT":      watchPort,
+		"OPENSHIFT_HA_REPLICA_COUNT":     replicas,
+		"OPENSHIFT_HA_USE_UNICAST":       unicastStr,
+		// "OPENSHIFT_HA_UNICAST_PEERS":     "127.0.0.1",
+	}
+}
+
+//  Generate the HA failover monitor (keepalived) container configuration.
+func generateFailoverMonitorContainerConfig(name string, options *haconfig.HAConfigCmdOptions, env app.Environment) *kapi.Container {
+	containerName := fmt.Sprintf("%s-%s", name, options.Type)
+
+	imageName := fmt.Sprintf("%s-%s", options.Type, haconfig.DefaultName)
+	image := options.ImageTemplate.ExpandOrDie(imageName)
+
+	ports := make([]kapi.ContainerPort, 1)
+	ports[0] = kapi.ContainerPort{
+		ContainerPort: options.ServicePort,
+		HostPort:      options.ServicePort,
+	}
+
+	mounts := make([]kapi.VolumeMount, 1)
+	mounts[0] = kapi.VolumeMount{
+		Name:      libModulesVolumeName,
+		ReadOnly:  true,
+		MountPath: libModulesPath,
+	}
+
+	return &kapi.Container{
+		Name:            containerName,
+		Image:           image,
+		Ports:           ports,
+		Privileged:      true,
+		ImagePullPolicy: kapi.PullIfNotPresent,
+		VolumeMounts:    mounts,
+		Env:             env.List(),
+	}
+}
+
+//  Generate the HA failover monitor (keepalived) container configuration.
+func generateContainerConfig(name string, options *haconfig.HAConfigCmdOptions) []kapi.Container {
+	containers := make([]kapi.Container, 0)
+
+	if len(options.VirtualIPs) < 1 {
+		return containers
+	}
+
+	config := getClientConfig(options.Credentials)
+	env := generateEnvEntries(name, options, config)
+
+	c := generateFailoverMonitorContainerConfig(name, options, env)
+	if c != nil {
+		containers = append(containers, *c)
+	}
+
+	return containers
+}
+
+//  Generate the HA failover monitor (keepalived) container volume config.
+func generateVolumeConfig() []kapi.Volume {
+	hostPath := &kapi.HostPathVolumeSource{Path: libModulesPath}
```

Comment:
Ok, add a comment here and make it clear this is why it's needed.

----- Original Message -----
> > +	}
> > +
> > +	config := getClientConfig(options.Credentials)
> > +	env := generateEnvEntries(name, options, config)
> > +
> > +	c := generateFailoverMonitorContainerConfig(name, options, env)
> > +	if c != nil {
> > +		containers = append(containers, *c)
> > +	}
> > +
> > +	return containers
> > +}
> > +
> > +//  Generate the HA failover monitor (keepalived) container volume config.
> > +func generateVolumeConfig() []kapi.Volume {
> > +	hostPath := &kapi.HostPathVolumeSource{Path: libModulesPath}
> 
> The keepalived sidecar container needs access to the kernel modules directory
> in order to load/unload it.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1807/files#r28801889